File: java/src/main/java/br/datastructures/stack/LinkedStack.java
package br.datastructures.stack;

import br.datastructures.exceptions.NoSuchItemException;

public class LinkedStack<I> extends AbstractStack<I> {
    class Node {
        I item;
        Node next;
        
        public Node() {
            this(null, null);
        }
        
        public Node(I item) {
            this(item, null);
        }
        
        public Node(I item, Node next) {
            this.item = item;
            this.next = next;
        }
    }
    
    private Node top;
    
    public LinkedStack() {
        top = null;
    }
    
    @Override
    public void push(I item){
        Node newNode = new Node(item, top);
        top = newNode;
        
        quantity++;
    }
    
    @Override
    public I pop() throws NoSuchItemException {
        I item = getTop();
        top = top.next;
        quantity--;

        return item;
    }
    
    @Override
    public I getTop() throws NoSuchItemException {
        if(empty()) 
            throw new NoSuchItemException();
        
        return top.item;
    }
}

File: c/stack/LinkedStack.c
#include <stdio.h>
#include <stdlib.h>

typedef int Item;

typedef struct No {
    Item item;
    struct No* next;
} No;

typedef struct {
    No* top;
    int quantity;
} LinkedStack;

LinkedStack* newLinkedStack() {
    LinkedStack* stack = (LinkedStack*) malloc(sizeof(LinkedStack));
    stack->top = NULL;
    stack->quantity = 0;
    
    return stack;
}

int tamanho(LinkedStack* stack) {
    return stack->quantity;
}

int vazia(LinkedStack* stack) {
    return tamanho(stack) == 0;
}

void push(LinkedStack* stack, Item item) {
    No* newNo = (No*) malloc(sizeof(No));
    newNo->item = item;
    newNo->next = stack->top;
    stack->top = newNo;

    stack->quantity++;
}

Item pop(LinkedStack* stack) {
    if(vazia(stack))
        return -1;

    Item item = getTop(stack);
    
    No* top = stack->top;
    stack->top = stack->top->next;

    free(top);

    stack->quantity--;

    return item;
}

Item getTop(LinkedStack* stack) {
    if(vazia(stack))
       return -1;
       
    return stack->top->item;
}

void viewStack(LinkedStack* stack) {
    while(!vazia(stack))
        printf("%d\n", pop(stack));

    free(stack);
}

void freeStack(LinkedStack* stack) {
    while(!vazia(stack))
        pop(stack);

    free(stack);
}

int main() {
    LinkedStack* stack = newLinkedStack();

    push(stack, 10);
    printf("%d\n", getTop(stack));

    push(stack, 11);
    printf("%d\n", getTop(stack));

    push(stack, 12);
    printf("%d\n", getTop(stack));

    printf("%d removido\n", pop(stack));

    viewStack(stack);

    return 0;
}